### ISSUE TITLE
CLI option that prevents equal validation metric from contributing to early stopping (related to --patience)

### ISSUE THREAD

**Issue by atreyasha** [pytorch/fairseq#2361]
## 🚀 Feature Request

This is a feature request to add an additional command-line option, for example `--tolerate-equal-validation`, that would prevent the patience counter from being incremented when the validation metric is reported to be exactly the same (ie. no explicit improvement or worsening of validation metric).

### Motivation

I was previously training a NMT model based off the Scaling NMT [readme](https://github.com/pytorch/fairseq/blob/master/examples/scaling_nmt/README.md) and I had used a `patience` value of 5. After a while, I noticed model training had stopped because the last 5 checkpoints had exactly the same validation metric; which was essentially the trigger.

While this behaviour conforms to the general idea behind `patience`, I was thinking it would have been useful to have had a feature that would not count equal validation performance towards the patience counter. This would have allowed the model to continue training and perhaps get past a plateau in performance; or otherwise go into a region of (explicitly) worse performance which would have triggered early stopping as well.

In my particular case, this behaviour would have been very helpful because I could have also used the (multiple) equally performing checkpoints (on the validation set) for checkpoint averaging later on, while still keeping `patience` to stop training when explicitly worse performance is observed.

### Pitch

With the addition of a (prospective)  `--tolerate-equal-validation` option, I could envision two possibilities:

1. Mild behaviour: the patience counter is not incremented if validation performance is exactly the same. The best checkpoint does not change.

2. Aggressive behaviour: the patience counter is set to zero if validation performance is exactly the same. The best checkpoint is also updated to the latest.

I am quite agnostic between the two, so would be great to get your opinions.

In either case, the relevant lines of code would be https://github.com/pytorch/fairseq/blob/master/fairseq_cli/train.py#L144, specifically:

```python
def is_better(a, b):
    return a > b if args.maximize_best_checkpoint_metric else a < b
```

I propose we could add an additional function there, such as:

```python
def is_better_or_equal(a, b):
    return a >= b if args.maximize_best_checkpoint_metric else a <= b
```

I would be happy to submit a PR if this is deemed of interest and/or value.

### Alternatives

An easy alternative could be to just use a higher `patience` value. But IMO this problem is relative in nature and could happen at any scale. This proposed feature would provide the user with an increased flexibility, which I believe corresponds to the general philosophy of the code-base of `fairseq`.
